airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -51,6 +52,7 @@ class _UpstreamTIStates(NamedTuple):
     done: int
     success_setup: int
     skipped_setup: int
+    failed_setup: int
```

Comment:
until now, everything in this class is *direct* upstreams. is failed_setup direct only or does it include indirect too? if it includes indirect too, it should probably be clarified through a more precise variable name.  but perhaps better would be to avoid mixing direct and indirect in the same class if it can be avoided.  perhaps we can just add the information through an optional argument in `calculate` or something.  this would also make the diff easier to deal with.

Reply:
the other issue is, there are more states than just "failed" that we need to account for.  there is also upstream_failed, skipped, etc -- essentially anything other than success.

Reply:
>  is failed_setup direct only or does it include indirect too?

yes it includes indirect too. The same for success_setup and skipped_setup

Reply:
For me there is 4 different cases:
- running (all the unfinished states) -> we should wait
- success -> we can run the task
- failed (failed or upstream_failed) -> fail
- skipped with any fail -> we should skipp
So failed here includes failed and upstream_failed:
```python
        return _UpstreamTIStates(
            success=counter.get(TaskInstanceState.SUCCESS, 0),
            skipped=counter.get(TaskInstanceState.SKIPPED, 0),
            failed=counter.get(TaskInstanceState.FAILED, 0),
            upstream_failed=counter.get(TaskInstanceState.UPSTREAM_FAILED, 0),
            removed=counter.get(TaskInstanceState.REMOVED, 0),
            done=sum(counter.values()),
            success_setup=setup_counter.get(TaskInstanceState.SUCCESS, 0),
            skipped_setup=setup_counter.get(TaskInstanceState.SKIPPED, 0),
            failed_setup=setup_counter.get(TaskInstanceState.FAILED, 0)
            + setup_counter.get(TaskInstanceState.UPSTREAM_FAILED, 0),
        )
```